File: p12/p12.cc
#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

vector<int> primes;

void GenPrimes() {
    primes.push_back(2);
    for (int i = 3; i < 1000000; i+=2) {
        for (auto j : primes) {
            if( i % j == 0) {
                break;
            }
            if( j > sqrt(i) ) {
                primes.push_back(i);
                break;
            }
        }
    }
}

int numDivisors(long input) {
    vector<int> nums;
    long old_input_sqrt = ceil(sqrt(input));
    for(auto p : primes) {
        if(p > old_input_sqrt) {
            if(nums.size() == 0) {
                return 2;
            }
            break;
        }
        int current_num = 0;
        while(input % p == 0) {
            current_num++;
            input /= p;
        }
        if(current_num > 0) {
            nums.push_back(current_num);
        }
    }
    int total = 1;
    for(auto num : nums) {
        total*= (num+1);
    }
    return total;
}

int main() {
    GenPrimes();
    long current = 1;
    long addr = 2;
    while(1) {
        current+=addr;
        addr++;
        if (numDivisors(current) > 500) {
            cout << current << endl;
            break;
        }
    }
}

File: p12/p12.py
import math

num_divisor_map = {}

primes = [2]
for i in range(1, 1000000):
  num = i*2 + 1
  for j in primes:
    if num % j == 0:
      break
    if j > math.sqrt(num):
      primes.append(num)
      break


def get_num_divisors(n):
  old_n = n
  total = 0
  for p in primes:
    if p > n:
      return total
    while n % p == 0:
      total += 1
      n = n/p
      if n == 1:
        if old_n == p:
          return total + 1
        return total + 2

get_num_divisors(16)


for i in range(1, 1000000):
  if get_num_divisors(i) > 5:
    print i
    break




# nth triangular number is 1 + 2 + 3 ... + n
# which is n(n + 1) / 2, or n+1 C 2
# n and n + 
